lmd/datarow.go

```diff
@@ -746,3 +746,184 @@ func (d *DataRow) WriteJSON(json *jsoniter.Stream, columns *[]*Column) {
+		contactsColumn := p.Tables["services"].GetColumn("contacts")
+		if ok {
+			contacts = append(contacts, *serviceObj.GetStringList(contactsColumn)...)
+		}
+	}
+
+	for _, contact := range contacts {
+		if contact == authUser {
+			canView = true
+			return
+		}
+	}
+
+	return
+}
+
+func (d *DataRow) isAuthorizedForGroup(authUser string, hostgroup string, servicegroup string) (canView bool) {
+	// Just use reflect for now, will be refactored when merging with upstream
+	// that has fixed typing.
+	var members []interface{}
+	p := d.DataStore.Peer
+	canView = false
+
+	if hostgroup != "" {
+		hostgroupObj, ok := p.Tables["hostgroups"].Index[hostgroup]
+		membersColumn := p.Tables["hostgroups"].GetColumn("members")
+		if ok {
+			// As the format for members on hostgroups differs from
+			// that of servicegroups, we manually add the items here.
+			stringMembers := *(*hostgroupObj).GetStringList(membersColumn)
+			for _, member := range stringMembers {
+				members = append(members, member)
+			}
+		}
+	} else if servicegroup != "" {
+		servicegroupObj, ok := p.Tables["servicegroups"].Index[servicegroup]
+		membersColumn := p.Tables["servicegroups"].GetColumn("members")
+		if ok {
+			members = (*servicegroupObj).GetInterfaceList(membersColumn)
+		}
+	}
+
+	for i := 0; i < len(members); i++ {
+		var hostName string
+		var serviceDescription string
+		if hostgroup != "" {
+			hostName = members[i].(string)
+		} else {
+			member := members[i].([]interface{})
+			hostName = member[0].(string)
+			serviceDescription = member[1].(string)
+		}
+		/* If GroupAuthorization is loose, we just need to find the contact
+		 * in any hosts in the group, then we can return.
+		 * If it is strict, the user must be a contact on every single host.
+		 * Therefore we return false if a contact doesn't exists on a host
+		 * and then on the last iteration return true if the contact is a contact
+		 * on the final host
+		 */
+		switch p.LocalConfig.GroupAuthorization {
+		case loose:
+			if d.isAuthorizedFor(authUser, hostName, serviceDescription) {
+				canView = true
+				return
+			}
+		case strict:
+			if !d.isAuthorizedFor(authUser, hostName, serviceDescription) {
+				canView = false
+				return
+			} else if i == len(members)-1 {
+				canView = true
+				return
+			}
+		}
+	}
+
+	return
+}
+
+func (d *DataRow) checkAuth(authUser string) (canView bool) {
+	canView = true
+	// authUser is only valid for the below tables:
+	tablesMap := map[string]struct{}{
+		"hosts":               {},
+		"services":            {},
+		"hostgroups":          {},
+		"servicegroups":       {},
+		"downtimes":           {},
+		"comments":            {},
+		"hostsbygroup":        {},
+		"servicesbygroup":     {},
+		"servicesbyhostgroup": {},
+	}
+	table := d.DataStore.Table
+	// check if this is a valid table and authUser is set
+	if _, ok := tablesMap[d.DataStore.Table.Name]; ok && authUser != "" {
+		canView = false
+		switch d.DataStore.Table.Name {
+		case "hosts", "services":
```

Comment:
this looks a bit awkward, first combine "hosts" and "services" in the same case, then again use `if` to do different things. I would simply use 2 cases here.